### PR TITLE
Fix GNO example

### PR DESCRIPTION
**Pull request by yuehhua**

### CODE REVIEW

**Comment by foldfelis** [comment]
Still need some revise, discussed F2F

**Comment by YichengDWu** [comment]
What is the loss after training?

**Comment by foldfelis** [comment]
@MilkshakeForReal Please take a look at [this](https://github.com/zongyi-li/fourier_neural_operator/blob/74b1572d4e02f215728b4aa5bf46374ed7daba06/utilities3.py#L161), it is the relative L2 loss

Update:

Sorry I misunderstand the question. If you are asking about the value of the loss after training, I'll get back to you later owing to the example is not implemented correctly yet

**Comment by YichengDWu** [comment]
Yes, I‘m asking about the value. The reason I'm asking is that I don't see any magical power in `GraphKernel` here. It is just an `NNConv` and could really be almost any GNNConvs.  And you see no justification for using it in the paper. The power of GNO is likely due to the encoder and sampling (or maybe something else?). Please let me know if the loss value is available, even if it's a large one.

**Comment by foldfelis** [comment]
I am not sure what kind of magic you expect to see in the GNO. Just like the eqs you mentioned in #74 is nothing but a message passing NN?

**Comment by foldfelis** [comment]
The main idea of neural operator is to learn the mapping in the spectral space. In GNO, the implementation is done by graph signal laplace transform and message passing neural network. @yuehhua is an expert of GNN, and maybe would like to explane more detail on this.

**Comment by yuehhua** [comment]
I think we all are doing science. There should be no magic.
All advantages GNO have over FNO is sampling. GNO doesn't require strictly grid sampling from input functions, but FNO do.
GNO should come out with at most the same performance as FNO do.

**Comment by YichengDWu** [comment]
When you expect to get GNO by just implementing the convolutional layer, you are expecting magic. As I already said, you must also implement the encoder and Nyström approximation.

**Comment by yuehhua** [comment]
Nyström approximation is already there and encoder is the `GraphKernel`. As for the graph convolutional layer, it is just a generalization of regular convolutional layer. The only magic thing should go to the non-linearity.

**Comment by YichengDWu** [comment]
Please read equation 7 carefully and make sure you understand what each item means

**Comment by yuehhua** [comment]
So, you mean the projection $v_0(x) = P(x, a(x), a_{\epsilon}(x), \nabla a_{\epsilon}(x)) + p$?

**Comment by foldfelis** [comment]
@MilkshakeForReal About the encoder, you mean [the `GaussianNormalizer`](https://github.com/zongyi-li/graph-pde/blob/08ab9d8f8e9d1a6d95017a6f803a53d8a603e66e/graph-neural-operator/utilities.py#L109)?

**Comment by YichengDWu** [comment]
Of course Nyström approximation is the more important thing. It is not clear where you have implemented it. I'll check back but now I really have to go to bed.

**Comment by foldfelis** [comment]
@MilkshakeForReal Take your time, and please feel free to open a PR if you still think there is anything wrong 😄

**Comment by foldfelis** [comment]
The $a_{\epsilon}(x)$ is the encoded $a(x)$ and the encoder is just a linear transform...

**Comment by YichengDWu** [comment]
Ok I'm back. In the spirit of scientific research, I don't want to discourage you from trying different things out. I do want to see the loss(es) first. But for now allow me to have a comment on Equation 7. 

No it's not the [GaussianNormalizer](https://github.com/zongyi-li/graph-pde/blob/08ab9d8f8e9d1a6d95017a6f803a53d8a603e66e/graph-neural-operator/utilities.py#L109). It normalizes the data, not smoothes it. The smoothed functions are already generated in the data, see [here](https://github.com/zongyi-li/graph-pde/blob/08ab9d8f8e9d1a6d95017a6f803a53d8a603e66e/graph-neural-operator/UAI1_full_resolution.py#L86). Unfortunately, there is no code to see how it is generated (or I missed that). The linear map `P` is less important I was not talking about that. Please note the motivation behind it
> Due to the smoothing effect of the inverse elliptic operator in (3) with respect to the input data a (and indeed f
when we consider this as input)

So the authors know how the solution operator acts on `a(x)`, and encode that info into the input. The model they actually use is restricted on the elliptic PDE `(4)`. And in the original paper they only test their model on this particular PDE :sweat_smile:. You can try removing $a_{\epsilon}(x)$ from the input and see how it affects the performance. It will show how general GNO actually is.

**Comment by YichengDWu** [comment]
> GNO doesn't require strictly grid sampling from input functions, but FNO do.

No FNO does not require that, only FFT does. General DFT can be performed on a nonuniform grid.

**Comment by yuehhua** [comment]
> No FNO does not require that, only FFT does. General DFT can be performed on a nonuniform grid.

Oh, yeah that's true. But could you give me nonuniform general DFT in practice?

**Comment by foldfelis** [comment]
@yuehhua Is this ready to merge?

**Comment by yuehhua** [comment]
@MilkshakeForReal For performing on a irregular domain, you may want to check [Geo-FNO](https://arxiv.org/abs/2207.05209).

@foldfelis CPU works, but not GPU. So, currently, not yet.

**Comment by YichengDWu** [comment]
Thanks for the info. Is Nyström approximation implemented in `GraphSignals.generate_grid`?

**Comment by yuehhua** [comment]
@MilkshakeForReal To my understand, there is no implementation for Nyström approximation, Nyström approximation is just a way to approximate kernel in RHKS. Thus, the model can really be computable, otherwise it is just an abstract mathematical concept. Please check paper in section 3 in p.8.

**Comment by YichengDWu** [comment]
I wouldn't need to, unless there is a new version just published. If you haven't implemented that's what I need to know.

**Comment by yuehhua** [comment]
In other words, `GraphKernel`, which is the GNN approximator itself, is the result of Nyström approximation of kernel for PDE.

**Comment by foldfelis** [comment]
@yuehhua  Great work, Tks

**Comment by foldfelis** [comment]
@yuehhua  Great work, Tks
